### DISCUSSION TITLE
[Question][Hw4]

### DISCUSSION THREAD

**Discussion by ZosiaZamoyska** [prosyslab-classroom/cs424-program-reasoning#343] [Q&A]
What is the meaning of "p' := p[B\A]" on the 13th slide [here](https://github.com/prosyslab-classroom/cs492-program-reasoning/blob/main/slides/lecture10.pdf)? I understood that part as an iteration of relations in production rules (grammar?), where there is a mapping from the type of A to something else (the result of Grammar.find?), and in "p' := p[B\A]" we should substitute one of the non-terminals of p for a terminal? 
What is the type of A and B?

**Comment by Lee-Janggun** [comment]
It means that one should replace the left most non-terminal `A` in `p` with the expanding rule given by `B`. For instance, consider the following term `p` and ruleset `R`
```
p = S + S
R := ( B -> B & B, B -> S = S, S -> S + S, S -> Ite B S S, S -> S - S, S -> x, S -> 0, S -> 1
```
Since the left most non-terminal of `p` is `S`, we will use the rules `S -> S + S`, `S -> Ite B S S`, `S -> 0` etc. After expanding, the following terms will be created.

```
(S + S) + S, (Ite B S S) + S, (S - S) + S, 0 + S, 1 + S
```

**Comment by KihongHeo** [comment]
Thanks Janggun. This is correct.
